Framework Agreement 2.0

lepistone opened the pull request:
Reverse dependencies:
- https://github.com/OCA/department/pull/39
- https://github.com/OCA/vertical-ngo/pull/128

lepistone committed:
delete unused file

lepistone committed:
bump: agreements two-dot-zero

lepistone committed:
suppress workflow warning, cleanup

lepistone committed:
refactor tests

I will adapt them to the new functionality, so I'm refactoring them to
help me understand them better.

lepistone committed:
bring back test we disabled

We disabled it because of odoo/odoo#3056, now fixed.

lepistone committed:
rename repetitive file names

lepistone committed:
refactor tests

I move some helper code closer to the only test that uses it.

lepistone committed:
remove on_change checks on the date

We will postpone consistency checks to the validation of the po.

lepistone committed:
add portfolio, remove some checks, keep tests green

yvaucher commented on framework_agreement/model/portfolio.py:
```diff
@@ -1,8 +1,5 @@
-#
-#    Author: Nicolas Bessi
-#    Copyright 2013, 2014 Camptocamp SA
+#    Author: Pistone
```
You forgot your firstname. :smile_cat:

nbessi commented:
hmmm, there is other addons that rely on framework agreement. Such changes will break other repositories (vertical ngo, etc) that may be a cause of opposition to PR if compatibility is not here.

lepistone committed:
lint, remove deprecated "version"

lepistone commented:
Thanks @nbessi. Indeed I need to make this compatible with the modules in OCA/vertical-ngo that depend on it. I will make the changes needed, and maybe revert something. This is early WIP, hold on!

guewen commented on framework_agreement/tests/test_consumed_qty.py:
```diff
@@ -0,0 +1,104 @@
+from .common import BaseAgreementTestMixin
+
+
+class TestAvailabeQty(test_common.TransactionCase, BaseAgreementTestMixin):
```
s/TestAvailabeQty/TestAvailableQty/

lepistone committed:
fix typo

lepistone committed:
portfolio, adapt views and tests

lepistone committed:
make portfolio unique per supplier-company

lepistone committed:
make related readonly, adapt to new company

lepistone committed:
fix: always ignore draft agreements

lepistone committed:
refactor overlap check

lepistone committed:
adapt overlap method

We now check on incoterms, as well.

lepistone committed:
portfolio and agreement on order line

lepistone committed:
mention myself as co-author

lepistone committed:
remove unused imports

lepistone committed:
new logic to choose the correct agreement

lepistone committed:
agreement required if portfolio, readonly if not

lepistone committed:
pull multi currency group

lepistone committed:
prevent to change supplier if portfolio is set

Before there was a similar check on the agreement.

lepistone committed:
refactor product onchange to reuse it further

lepistone committed:
update agreement in lines when the header changes

lepistone committed:
lint

lepistone committed:
move get_agreement_domain method to agreement

This is because it is actually not specific to the purchase order. I
will use it again in logistic sources.

lepistone committed:
in view, refer only new-style onchanges

lepistone committed:
update doc, remove outdated sentence

lepistone committed:
accept incoterm address, optional portfolio

Since the incoterm_address here is a field with default None, it is not
a problem that the field is not introduced in this module

nbessi commented:
@lepistone can you link other PR related to this in other project please

lepistone committed:
fix error when you remove the portfolio

lepistone committed:
add missing quantity onchange

yvaucher commented on framework_agreement/__openerp__.py:
```diff
@@ -29,7 +27,7 @@
 Long Term Agreement (or Framework Agreement) on price.
```
Can we put this in README.rst ?

yvaucher commented on framework_agreement/__openerp__.py:
```diff
@@ -19,7 +17,7 @@
 #
 ##############################################################################
 {'name': 'Framework Agreement',
```
Suggestion:
`'summary': 'Long term agreement for purchases',`

lepistone replied:
done

yvaucher commented on framework_agreement/model/framework_agreement.py:
```diff
@@ -311,72 +315,57 @@ def create(self, vals):
 
-        One agreement per product limit is checked if one_agreement_per_product
-        is set to True on company
+        Depending on the  one_agreement_per_product flag on the company,
```
Extra space before flag name

lepistone committed:
update doc

lepistone committed:
lint

yvaucher commented on framework_agreement/model/framework_agreement.py:
```diff
@@ -268,10 +272,10 @@ def _compute_state(self):
+                 'purchase_line_ids.framework_agreement_id',
+                 'purchase_line_ids.order_id.state',
+                 'purchase_line_ids',
+                 'purchase_line_ids.product_qty')
```
Those triggers seems strange compared to `_compute_state` method.

Aren't `state`, `start_date`, `end_date` missing?

About computed fields `available_quantity` I guess it doesn't aggregate dependencies of dependencies? You could add a comment to explain why we keep all those dependences?
like `# dependencies of available_quantity`

lepistone replied:
ok, done

yvaucher replied:
thanks

yvaucher commented on framework_agreement/model/framework_agreement.py:
```diff
@@ -311,72 +315,57 @@ def create(self, vals):
+        )
+
+        if self.company_id.one_agreement_per_product:
+            # in stict mode, any overlap is bad
```
s/stict/strict

yvaucher commented on framework_agreement/model/framework_agreement.py:
```diff
@@ -403,6 +392,12 @@ def get_all_product_agreements(self, product_id, lookup_dt, qty=None):
 
+    # probably unused
+    # @api.multi
+    # def get_cheapest_in_set(self, qty, currency):
```
This is not used in OCA addons you can remove it

yvaucher commented on framework_agreement/model/purchase.py:
```diff
@@ -1,8 +1,5 @@
-# -*- coding: utf-8 -*-
```
Don't remove this line please or someone with an accent in his name couldn't contribute to that file

lepistone replied:
My reasoning was that this is not dangerous because in python 2 the default is ASCII so that if Amélie or Joël contribute to that file, their name will not be garbled (which is bad), but they will get a nice error explaining what to do.

I like to get rid of stuff if the default is good enough, but will put that back if you insist :cat:

yvaucher replied:
I think we should encourage contribution of anyone. Thus I prefer to keep it.

lepistone replied:
The little kitten didn't work, so I'm putting that back.

yvaucher replied:
:smiling_imp:

yvaucher commented on framework_agreement/model/purchase.py:
```diff
@@ -191,20 +188,13 @@ def _propagate_fields(self):
-                      'Agreement and purchase date does not match')
-                )
-            if agreement.supplier_id.id != self.partner_id:
+            if agreement.supplier_id.id != self.order_id.partner_id:
```
wrong comparison id on left and record set on right

lepistone committed:
kill dead code

lepistone committed:
fix typo

lepistone committed:
fix comparison of apples and pears

lepistone committed:
add comment on dependency

yvaucher commented:
:+1: Great job for this new version

lepistone committed:
revert to old view names to fix crash on update

lepistone committed:
show supplier in agreement view

lepistone committed:
improve error message

lepistone committed:
fix predicate on api.one: was always true

That happens because api.one always encloses the results in a list, and
bool([False]) = True

lepistone committed:
make onchange agreement work on new order

lepistone committed:
update the price when an agreement is chosen

lepistone committed:
fix hideous indentation bug

lepistone committed:
add migration script to create missing portfolios

gurneyalex commented on framework_agreement/model/framework_agreement.py:
```diff
@@ -311,72 +316,56 @@ def create(self, vals):
+            # includes the current self
+            ('start_date', '<=', self.start_date),
+            ('end_date', '>=', self.end_date),
+        ]
```
if you have a constraint that end_date > start_date, then you can remove line 341-343 (and line 330).

gurneyalex commented on framework_agreement/model/framework_agreement.py:
```diff
@@ -311,72 +316,56 @@ def create(self, vals):
+            # in non-strict mode, same-supplier overlap is bad
+            if any(o.supplier_id == self.supplier_id for o in overlap):
+                raise exceptions.ValidationError(
+                    _('There can only be one agreement for every combination '
```
s/for every combination/for a given supplier, incoterm, incoterm address and product/

gurneyalex commented:
minor changes to be made, but most importantly, fix the build of https://github.com/OCA/vertical-ngo/pull/128

lepistone committed:
new method to get portfolio from supplier

lepistone committed:
simplify the overlap domain

lepistone committed:
rephrase constraint message
